streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -748,24 +702,25 @@ private void assignTasksToClients(final Set<String> allSourceTopics,
         final boolean lagComputationSuccessful =
             populateClientStatesMap(clientStates, clientMetadataMap, taskForPartition, changelogsByStatefulTask);
 
-        // assign tasks to clients
         final Set<TaskId> allTasks = partitionsForTask.keySet();
-        final Set<TaskId> standbyTasks = changelogsByStatefulTask.keySet();
-
-        if (lagComputationSuccessful) {
-            final Map<TaskId, SortedSet<RankedClient<UUID>>> statefulTasksToRankedCandidates =
-                buildClientRankingsByTask(standbyTasks, clientStates, acceptableRecoveryLag());
-            log.trace("Computed statefulTasksToRankedCandidates map as {}", statefulTasksToRankedCandidates);
-        }
+        final Set<TaskId> statefulTasks = changelogsByStatefulTask.keySet();
 
         log.debug("Assigning tasks {} to clients {} with number of replicas {}",
             allTasks, clientStates, numStandbyReplicas());
 
-        final StickyTaskAssignor<UUID> taskAssignor = new StickyTaskAssignor<>(clientStates, allTasks, standbyTasks);
-        if (!lagComputationSuccessful) {
-            taskAssignor.preservePreviousTaskAssignment();
+        final TaskAssignor<UUID> taskAssignor;
+        if (highAvailabilityEnabled) {
+            if (lagComputationSuccessful) {
+                taskAssignor = new HighAvailabilityTaskAssignor<>(clientStates, allTasks, statefulTasks,
+                    assignmentConfigs);
+            } else {
+                taskAssignor = new StickyTaskAssignor<>(clientStates, allTasks, statefulTasks, assignmentConfigs);
+                ((StickyTaskAssignor) taskAssignor).preservePreviousTaskAssignment();
+            }
+        } else {
+            taskAssignor = new StickyTaskAssignor<>(clientStates, allTasks, statefulTasks, assignmentConfigs);
```

Comment:
Thanks, this block looks more understandable to me now.

New question: previously, we would set the sticky assignor to preserve previous task assignments if lag computation weren't successful, but now we only do that if we're falling back from the HA task assignor. Did you mean to also include the "if (!lagComputationSuccessful){ preserve previous }` logic here?

It seems a little hairy, and it does seem clear that we should refactor it, but I'm ok with doing that in a follow-up PR, if you are.

Reply:
I was referring to "prior" as in the starting point of this diff, aka current trunk.

I think it has the `lagComputationSuccessful` check there because of the way we're inferring active/standby status based on the reported lag, which is not correct if we couldn't actually compute the lags, so we put that "safety valve" in to just no-op the assignment and try again next time.

Reply:
Just to close the loop for anyone coming back to read this, the `lagComputationSuccessful` only fails if the end offset fetch fails, ie we will still have the offset sums. We can infer the previous task sets from the offset sums, so even if the lag computation fails, the `StickyTaskAssignor` can do its thing.
@vvcephei would it have helped to rename `lagComputationSuccessful` --> `endOffsetFetchSuccessful` or the like? Or was it mostly just the partial-KIP-441-ness of the current trunk that threw you off